# Convolutional Neural Network

**Name - Mitul Srivastava**

**ID - C00313606**


## **LOG** : Introduction to dataset and importing the data
### **DATASET** : CIFAR-10 dataset
### **DETAIL** : The CIFAR-10 dataset contains 60,000 color images (32x32) in 10 classes, with 50,000 training images and 10,000 test images.
### **AIM** : To train and fine tune Convolutional Neural network to correctly identify labels for images. 

## **LOG** : Importing required packages

In [3]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt

## **LOG** : Loading CIFAR-10 dataset using keras.datasets

In [4]:
cifar10 = keras.datasets.cifar10
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 48s 0us/step


## **LOG** : Normalising the dataset so that max value is 1 and min value is 0 by dividing by 255 because pixel has values from 0 to 255.
## This type of normalisation is called min-max scaling.

In [5]:
x_train, x_test = x_train / 255.0, x_test / 255.0

## **LOG** : Define the CNN model -
## 1. Conv2D: Extracts features from images using filters.
## 2. MaxPooling2D: Reduces spatial dimensions to prevent overfitting.
## 3. Flatten: Converts 2D feature maps into 1D.
## 4. Dense: Fully connected layers for classification.

In [6]:
model = keras.Sequential([
    keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(32,32,3)),
    keras.layers.MaxPooling2D(2,2),
    keras.layers.Conv2D(64, (3,3), activation='relu'),
    keras.layers.MaxPooling2D(2,2),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

C:\Users\Mitul\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


## **LOG** : Compiling the model with optimizer as "adam" and using "sparse_categorical_crossentropy" suitable for multi class classification for error and accuracy as the metric.

In [7]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

## **LOG** : Training the model with 10 epochs.

In [8]:
model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 26s 16ms/step - accuracy: 0.3813 - loss: 1.6903 - val_accuracy: 0.5721 - val_loss: 1.2244
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 23s 15ms/step - accuracy: 0.6008 - loss: 1.1382 - val_accuracy: 0.6396 - val_loss: 1.0355
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.6668 - loss: 0.9596 - val_accuracy: 0.6657 - val_loss: 0.9658
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.7026 - loss: 0.8549 - val_accuracy: 0.6841 - val_loss: 0.9253
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.7371 - loss: 0.7626 - val_accuracy: 0.6840 - val_loss: 0.9282
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.7610 - loss: 0.6964 - val_accuracy: 0.7021 - val_loss: 0.8841
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.7809 - loss: 0.6336 - val_accuracy: 0.7061 - val_loss: 0.9067
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 22s 14ms/step - accuracy: 0.8043 -

## **LOG** : Model evaluation

In [9]:
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc:.4f}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.6993 - loss: 0.9887
Test accuracy: 0.6973


## **LOG**: Improving the performance by -
## 1. Increase filters in Conv2D layers to capture more features.
## 2. Add Dropout layers to reduce overfitting.
## 3. Increasing epochs to 15.

In [10]:
model_improved = keras.Sequential([
    keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(32,32,3)),
    keras.layers.MaxPooling2D(2,2),
    keras.layers.Conv2D(128, (3,3), activation='relu'),
    keras.layers.MaxPooling2D(2,2),
    keras.layers.Flatten(),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(10, activation='softmax')
])

## **LOG** : Compiling the improved model.

In [11]:
model_improved.compile(optimizer='adam',
                        loss='sparse_categorical_crossentropy',
                        metrics=['accuracy'])

model_improved.fit(x_train, y_train, epochs=15, validation_data=(x_test, y_test))

Epoch 1/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 64s 40ms/step - accuracy: 0.3812 - loss: 1.6870 - val_accuracy: 0.5856 - val_loss: 1.1638
Epoch 2/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 65s 42ms/step - accuracy: 0.6089 - loss: 1.1174 - val_accuracy: 0.6512 - val_loss: 1.0227
Epoch 3/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 64s 41ms/step - accuracy: 0.6745 - loss: 0.9407 - val_accuracy: 0.6906 - val_loss: 0.9021
Epoch 4/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 64s 41ms/step - accuracy: 0.7109 - loss: 0.8304 - val_accuracy: 0.7087 - val_loss: 0.8516
Epoch 5/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 65s 42ms/step - accuracy: 0.7404 - loss: 0.7400 - val_accuracy: 0.7015 - val_loss: 0.8773
Epoch 6/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 65s 41ms/step - accuracy: 0.7721 - loss: 0.6527 - val_accuracy: 0.7206 - val_loss: 0.8254
Epoch 7/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 65s 42ms/step - accuracy: 0.7914 - loss: 0.5873 - val_accuracy: 0.7159 - val_loss: 0.8581
Epoch 8/15
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 64s 41ms/step - accuracy: 0.8162 -

## **LOG** : Evaluating the improved model.
## We observe that the after modification the new models accuracy increased from 0.6973 to 0.7160.

In [12]:
test_loss, test_acc = model_improved.evaluate(x_test, y_test)
print(f"Improved Test accuracy: {test_acc:.4f}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 3s 10ms/step - accuracy: 0.7223 - loss: 1.0503
Improved Test accuracy: 0.7160


### **REFERENCES** :
### https://chatgpt.com/
### https://www.perplexity.ai/
### https://github.com/trekhleb/machine-learning-experiments

## **END**